### PR TITLE
Update to C# 11

### PR DESCRIPTION
**Pull request by jnyrup**
## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [ ] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [ ] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [ ] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

Updates to C# 11 and switches to list patterns a few places.

### CODE REVIEW

**Review by dennisdoomen — APPROVED** [review]
It's definitely a nice feature, but I wonder how it affects performance.

File: Src/FluentAssertions/CallerIdentification/MultiLineCommentParsingStrategy.cs [hunk 1]
```diff
@@ -26,10 +26,9 @@ public ParsingState Parse(char symbol, StringBuilder statement)
             return ParsingState.GoToNextSymbol;
         }
 
-        var isStartOfMultilineComment =
-            symbol is '*'
-            && statement.Length > 0
-            && statement[^1] is '/';
+#pragma warning disable SA1010 // https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/3507
+        var isStartOfMultilineComment = symbol is '*' && statement is [.., '/'];
+#pragma warning restore SA1010
```
Comment (jnyrup) [inline Src/FluentAssertions/CallerIdentification/MultiLineCommentParsingStrategy.cs:31]: Difference: 
A null check of `statement`

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRgDoAVGBAFwG4BYAKDuAggBscAFAQygGcYB5ZgBMAFGAAWXHNwCeAW0bMANDgDK5KAEsAdgHMAQgFcNQmFCnkO5GLJhbyASjoBvOjjc4AbpI3c1XcnwAZgCyBszkxtowAMIQsjZ2OAC8ru5paTLyLDg+OADkAFR5qenpAGRl5pbWtuSEADK2OuRiOAB8OFglpe4VVVYJ5ADaAHr4ALo53PkA9Hk0tGndGADsU35QASFhEcxRsfG1CwC+dAxMrJw8MAByMADuohJmmQrKapq6hsaCpv01dkctBci3cXjMPg2W1C4UiWhicUGySkcgUU3yRRwfW4FgGtXRQ0IhGUeTm4wWS1BblW6wsmyCMN2+0RRzoxyAA=)

image

File: Src/FluentAssertions/Common/ExpressionExtensions.cs [hunk 2]
```diff
@@ -87,15 +87,15 @@ internal static class ExpressionExtensions
                 case ExpressionType.Call:
                     var methodCallExpression = (MethodCallExpression)node;
 
-                    if (methodCallExpression.Method.Name != "get_Item" || methodCallExpression.Arguments.Count != 1 ||
-                        methodCallExpression.Arguments[0] is not ConstantExpression)
+#pragma warning disable SA1010 // https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/3507
+                    if (methodCallExpression is not { Method.Name: "get_Item", Arguments: [ConstantExpression argumentExpression] })
```
Comment (jnyrup) [inline Src/FluentAssertions/Common/ExpressionExtensions.cs:91]: Difference:
Null checks on `methodCallExpression`, `methodCallExpression.Method` and `methodCallExpression.Arguments`

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4B0AMgJYB2AjpQKIIAOsAzl7RPVwDchQhSSkIwAFYwwAF1IA1AIYAbWgBNlcmAFkYAW2AwoABW0ALAHIwA7gB4AKgBFZq5VAYBzRwE92MAA0pI6mUBABUHK+fgEAfAAUbJwwPHz09hgArE6uYO6e9D7+QSFhESbRsTBxcaQwHNy8/ACUhADehKTdpMlN6aT0EBowpAC89Y2pzfTCBD2kXLa0cmAWpAlDI5RWwzDVbfM9nUcL3WDKXKN90+nVlADCaqogXWfdAG4epAYwchbDJ6qVQ3NL8cYbfT/QHPUEzFpbGBzN7vWgAMw2v2hGiBIKmYPopFoXEGEAU7VIUIBGh2yl+IFIACIvH8APoASR0BkZwQAglAvABXX70ORcBkAbQe/C4cmUorhAw8QpFckV/AAuqQAL6Hd7dE76s7/cK2QZ2Uj8lUwBUIMAwdhydIJFpzI3alFnT0LDAAdlIyuFNrV+JmlBUqkFSM9HtOpE9fsGguBc1johw4kkMnkSjUmm0ekMxjMlgA8qoNLk3B5vNVgqFwpEqiVEuqMtkq/ka0U62VG5UYi26g0UgTDoaem3SSMISP+vw3T0lis1htETs9gdPRP3hcrr1Q3cSo9nq84wsvlAfn9qbipxMElSYcC2wi9sjzz10Zib8+8aOwyfGkrDpUYAEIJmZNlOUMRlSDQNBr2xO9D34SgrSDUUuEeCBBVFUgINIHB4LQb0jXI7osVvWFUPodCBUwsUJWILViVJBRpQEOVbQA9I9X1HdyJNCAzXoC0MNVNh7UdZ1XTI7pYwo7p5NIS9SEgLj5RDXjwQfTjZS018qL/Nt6OtLDmI1D8jUTDSDJ4+c6IjKNFwUz0E39ehk1UVz4wIbUgA)

image

**Review by jnyrup — COMMENTED** [review]
> It's definitely a nice feature, but I wonder how it affects performance.

Performance wise it's equivalent.
The only difference is a few added null checks that are technically superfluous.

List patterns is just a language construct that is lowered to regular C# code.
The only pitfall I know of is when extracting a sub-sequence of a non-span, then a _new_ sub-array is created. 
[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQhGKABAGIQQAUAlgHYAuA2gLqm0CUhA3oaQNIB9UgF4OHAM6lmAOlmkAbgEMopABYxlAEwA0pAAysA3IQC+hYmUoQs1AMoAHZfQA8DRgD4O3AnwKCwmIStNJyCipqmjr6RqYEZkA)

image

**Pull request merged**